src/puppetlabs/puppetdb/query/catalogs.clj

```diff
@@ -1,106 +1,111 @@
-  (:require [puppetlabs.puppetdb.schema :as pls]
-            [puppetlabs.puppetdb.catalogs :as cats]
-            [schema.core :as s]
-            [puppetlabs.puppetdb.utils :as utils]
-            [puppetlabs.puppetdb.query :as query]
+  "Catalog retrieval"
+  (:require [clojure.set :as set]
+            [puppetlabs.kitchensink.core :as kitchensink]
+            [puppetlabs.puppetdb.catalogs :as catalogs]
             [puppetlabs.puppetdb.cheshire :as json]
             [puppetlabs.puppetdb.jdbc :as jdbc]
+            [puppetlabs.puppetdb.query :as query]
             [puppetlabs.puppetdb.query.paging :as paging]
             [puppetlabs.puppetdb.query-eng.engine :as qe]
-            [puppetlabs.kitchensink.core :as kitchensink]))
+            [puppetlabs.puppetdb.schema :as pls]
+            [puppetlabs.puppetdb.utils :as utils]
+            [schema.core :as s])
+  (:import  [org.postgresql.util PGobject]))
 
-;; v4+ functions
-
-(def catalog-columns
-  [:certname
-   :version
-   :transaction_uuid
-   :producer_timestamp
-   :environment
-   :hash
-   :edges
-   :resources])
+;; SCHEMA
 
 (def row-schema
-  {:version (s/maybe String)
-   :hash (s/maybe String)
-   :transaction_uuid (s/maybe String)
-   :environment (s/maybe String)
-   :certname String
+  "Catalog query row schema."
+  {:certname s/Str
+   :edges (s/maybe PGobject)
+   :environment (s/maybe s/Str)
+   :hash (s/maybe s/Str)
    :producer_timestamp (s/maybe pls/Timestamp)
-   :resource (s/maybe String)
-   :type (s/maybe String)
-   :title (s/maybe String)
-   :tags (s/maybe [String])
-   :exported (s/maybe s/Bool)
-   :file (s/maybe String)
-   :line (s/maybe s/Int)
-   :parameters (s/maybe String)
-   :source_type (s/maybe String)
-   :source_title (s/maybe String)
-   :target_type (s/maybe String)
-   :target_title (s/maybe String)
-   :relationship (s/maybe String)})
+   :resources (s/maybe PGobject)
+   :transaction_uuid (s/maybe s/Str)
+   :version (s/maybe s/Str)})
 
-(defn catalog-response-schema
-  "Returns the correct schema for the `version`, use :all for the full-catalog (superset)"
-  [api-version]
-  (case api-version
-    :v4 (assoc (cats/catalog-wireformat :v5) :hash String)
-    (cats/catalog-wireformat api-version)))
+;; MUNGE
 
-(defn create-catalog-pred
-  [rows]
-  (let [catalog-hash (:hash (first rows))]
-    (fn [row]
-      (= catalog-hash (:hash row)))))
+(pls/defn-validated rtj->resource :- catalogs/resource-query-schema
+  "Convert the row_to_json PG output to real data, and parse parameters
+  from its JSON storage."
+  [resource :- {s/Str s/Any}]
+  (-> resource
+      (set/rename-keys {"f1" :resource
+                        "f2" :type
+                        "f3" :title
+                        "f4" :tags
+                        "f5" :exported
+                        "f6" :file
+                        "f7" :line
+                        "f8" :parameters})
+      (update-in [:parameters] #(json/parse-strict-string % true))))
```

Comment:
rename-keys here is not nearly as costly as for edges, but using set/rename instead of mapping this function may still be a good idea.